### PR TITLE
dd: open stdin from file descriptor when possible

### LINKED ISSUE #3008: dd: unexpectedly consumes all bytes from stdin
Executive summary:

Setup:
```
cargo build -p uu_dd
echo 'abcd' > infile
```

GNU dd:
```
$ (dd bs=1 skip=1 count=0 status=none && cat -) < infile
bcd
```

uutils dd:
```
$ (./target/debug/dd bs=1 skip=1 count=0 status=none && cat -) < infile
```
(it produces no output).

-----

I'm going to be very explicit about this because it helped me understand the problem here. The way I'm reading the command line is:
* `dd` means copy bytes from stdin to stdout,
* `bs=1` means blocks of 1 byte from the input and write blocks of 1 byte to the output,
* `skip=1` means skip 1 block of the input,
* `count=0` means write 0 blocks to the output,
* `&&` means do the second command only if the first command terminates with exit status 0,
* `cat -` means copy bytes from stdin to stdout,
* `< infile` means redirect the bytes of `infile` into stdin,

Altogether this could be described in English as "send bytes of `infile` into stdin, have `dd` skip the first byte and stop, then have `cat` copy the remaining bytes to stdout".

So the problem seems to be that by the time `cat` attempts to read from `stdin`, there are no bytes available to read, but there should be. I took a look in the code for `dd` and it doesn't seem to be reading more bytes than it should: https://github.com/uutils/coreutils/blob/1194a8ce534e88cf213ddc3b66bb8d357455cf56/src/uu/dd/src/dd.rs#L182 Perhaps this has something to do with how Rust provides a handle to stdin?

(I discovered this issue when looking into test failures in the GNU test `tests/misc/head-c.sh`.)

### PR DESCRIPTION
**Pull request by jfinkels**
~This is a draft pull request because it includes the changes from pull request #4164. Review and merge that one first.~

This pull request causes `dd` to open stdin using its file descriptor so that a `dd skip=N` command in a subshell does not consume all bytes from stdin.

For example, before this commit, multiple instances of `dd` reading from stdin and appearing in a single command line would incorrectly result in an empty stdin for each instance of `dd` after the first:

    $ printf "abcdef\n" | (dd bs=1 skip=3 count=0 && dd)  2> /dev/null
    # incorrectly results in no output

After this commit, the `dd skip=3` process reads three bytes from the file descriptor referring to stdin without draining the remaining three bytes when it terminates:

    $ printf "abcdef\n" | (dd bs=1 skip=3 count=0 && dd)  2> /dev/null
    def

This implementation is inspired by #3662. I don't think this quite resolves #3008 but it is in the right direction. It does seem to get GNU test `tests/dd/skip-seek2.sh` to pass.

### CODE REVIEW

**Comment by jfinkels** [comment]
I rebased this on `main` now that #4164 has been merged.

**Comment by tertsdiepraam** [comment]
Could you add a test for this somehow? Maybe by testing the script from the description?

**Comment by jfinkels** [comment]
I'll give it a shot and report back.

**Comment by tertsdiepraam** [comment]
Not sure, but this might help: https://github.com/uutils/coreutils/pull/4293

**Comment by jfinkels** [comment]
@Joining7943 since you worked on the `UCommand` implementation recently in #4293, do you know if it provides an obvious way for us to create a test like the following, where there are two invocations of the `dd` util joined by an `&&`:
```
$ printf "abcdef\n" | (dd bs=1 skip=3 count=0 && dd)  2> /dev/null
def
```

**Comment by Joining7943** [comment]
I agree with @tertsdiepraam. The most obvious way I can think of, is to use the shell, although I don't know if the windows `cmd` shell supports the parentheses the same way the `sh` shell does. Maybe somethink like this?

```rust
use crate::common::util::; // includes the constant TESTS_BINARY

#[cfg(feature = "printf")
#[test]
fn test() {
UCommand::new().arg(format!("{TESTS_BINARY} printf 'abcdef\n' | ( {TESTS_BINARY} dd bs=1 skip=3 count=0 && {TESTS_BINARY} dd ) 2> /dev/null")).run();

// or if you need to pass different arguments to the shell for unix and windows

UCommand::new().arg(
  #[cfg(unix)]
  // unix arg
  #[cfg(windows)]
  // windows arg
).run();
}
```

**Comment by jfinkels** [comment]
I'll try that. Thank you!

**Comment by Joining7943** [comment]
Sure :)

**Comment by jfinkels** [comment]
Rebased on main and added a test.

**Comment by jfinkels** [comment]
The remaining CI error is due to #4487. Other than that, this is ready for review again

File: src/uu/dd/src/dd.rs [hunk 1]
```diff
@@ -116,6 +142,15 @@ impl Source {
                 Ok(m) => Ok(m),
                 Err(e) => Err(e),
             },
+            #[cfg(unix)]
+            Self::StdinFile(f) => match io::copy(&mut f.take(n), &mut io::sink()) {
```
Comment (tertsdiepraam) [inline src/uu/dd/src/dd.rs:146]: I suppose we can't use file seeking operations here, because stdin doesn't support that even when you open as file?

Comment (jfinkels) [reply src/uu/dd/src/dd.rs:146]: I didn't try, but that was my assumption.

File: src/uu/dd/src/dd.rs [hunk 2]
```diff
@@ -116,6 +142,15 @@ impl Source {
...
+                Ok(m) if m < n => {
+                    show_error!("'standard input': cannot skip to specified offset");
```
[the remark this answers is not in the document]
Comment (jfinkels) [reply src/uu/dd/src/dd.rs:148]: It seems not. See the test case `test_skip_beyond_file` (which is misnamed, it is testing for skipping beyond stdin). This command has exit status 0 even though the highlighted error message is printed:
```
printf "abcd" | dd bs=1 skip=5 count=0
```

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/dd/src/dd.rs:148]: Alright!

File: src/uu/dd/src/dd.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uu/dd/src/dd.rs:146]: Is it worth trying? I'm fine with the current implementation too.

Comment (jfinkels) [reply src/uu/dd/src/dd.rs:146]: I can take a look now and remind myself what happens when using `seek()`. I'll report back

Comment (jfinkels) [reply src/uu/dd/src/dd.rs:146]: It seems to cause a bunch of `Broken pipe (os error 32)` and `Illegal seek` errors throughout the tests if I use `f.seek()` here.

Comment (tertsdiepraam) [reply src/uu/dd/src/dd.rs:146]: Alright, thanks!

**Pull request merged**